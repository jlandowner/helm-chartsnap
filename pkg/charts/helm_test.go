package charts

import (
	"context"

	. "github.com/jlandowner/helm-chartsnap/pkg/snap/gomega"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

var _ = Describe("Helm", func() {
	Context("when Execute", func() {
		It("should execute with expected args and env", func() {
			o := &HelmTemplateCmdOptions{
				HelmPath:    "./testdata/helm_cmd.bash",
				ReleaseName: "aaa",
				Namespace:   "bbb",
				Chart:       "ccc",
				ValuesFile:  "ddd",
			}

			out, err := o.Execute(context.Background())
			Expect(err).NotTo(HaveOccurred())
			Expect(out).To(MatchSnapShot())
		})
	})

	Context("when Execute without namespace", func() {
		It("should execute with expected args and env", func() {
			o := &HelmTemplateCmdOptions{
				HelmPath:    "./testdata/helm_cmd.bash",
				ReleaseName: "chartsnap",
				Chart:       "charts/app1/",
				ValuesFile:  "charts/app1/test/test.values.yaml",
			}

			out, err := o.Execute(context.Background())
			Expect(err).NotTo(HaveOccurred())
			Expect(out).To(MatchSnapShot())
		})
	})

	Context("when Execute without values", func() {
		It("should execute with expected args and env", func() {
			o := &HelmTemplateCmdOptions{
				HelmPath:    "./testdata/helm_cmd.bash",
				ReleaseName: "chartsnap",
				Namespace:   "default",
				Chart:       "charts/app1/",
			}

			out, err := o.Execute(context.Background())
			Expect(err).NotTo(HaveOccurred())
			Expect(out).To(MatchSnapShot())
		})
	})

	Context("when Execute with additional args", func() {
		It("should execute with expected args and env", func() {
			o := &HelmTemplateCmdOptions{
				HelmPath:       "./testdata/helm_cmd.bash",
				ReleaseName:    "chartsnap",
				Namespace:      "xxx",
				Chart:          "postgres",
				ValuesFile:     "postgres.values.yaml",
				AdditionalArgs: []string{"--repo", "https://charts.bitnami.com/bitnami", "--skip-tests"},
			}

			out, err := o.Execute(context.Background())
			Expect(err).NotTo(HaveOccurred())
			Expect(out).To(MatchSnapShot())
		})
	})

	Context("when Execute with post-renderer output containing multiline strings", func() {
		It("should execute with expected output", func() {
			o := &HelmTemplateCmdOptions{
				HelmPath:       "./testdata/helm_post_renderer.bash",
				ReleaseName:    "test",
				Chart:          "test-chart",
				AdditionalArgs: []string{"--post-renderer", "./fixTemplatePatch.sh"},
			}

			out, err := o.Execute(context.Background())
			Expect(err).NotTo(HaveOccurred())
			Expect(out).To(MatchSnapShot())
		})
	})

	Context("test mocks", func() {
		It("should execute as helm cmd", func() {
			o := &HelmTemplateCmdOptions{
				HelmPath:    "./testdata/helm_empty.bash",
				ReleaseName: "release",
				Chart:       "chart",
			}
			out, err := o.Execute(context.Background())
			Expect(err).NotTo(HaveOccurred())
			Expect(out).To(MatchSnapShot())
		})
	})
})
