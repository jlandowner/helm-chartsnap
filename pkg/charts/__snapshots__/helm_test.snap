['Helm test mocks should execute as helm cmd 1']
SnapShot = ''

['Helm when Execute should execute with expected args and env 1']
SnapShot = """
Arguments for helm: template aaa ccc --namespace=bbb --values=ddd
Environment variables starting with HELM_:
HELM_DEBUG=false
"""

['Helm when Execute with additional args should execute with expected args and env 1']
SnapShot = """
Arguments for helm: template chartsnap postgres --namespace=xxx --values=postgres.values.yaml --repo https://charts.bitnami.com/bitnami --skip-tests
Environment variables starting with HELM_:
HELM_DEBUG=false
"""

['Helm when Execute with post-renderer output containing multiline strings should execute with expected output 1']
SnapShot = """
---
# Source: test-chart/templates/secret.yaml
apiVersion: v1
kind: Secret
metadata:
  name: test-secret
stringData:
  certificate: \"-----BEGIN CERTIFICATE-----
MIIDXTCCAkWgAwIBAgIJAKZ
-----END CERTIFICATE-----\"
---
# Source: test-chart/templates/configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: test-config
data:
  message: \"hello world\"
"""

['Helm when Execute without namespace should execute with expected args and env 1']
SnapShot = """
Arguments for helm: template chartsnap charts/app1/ --values=charts/app1/test/test.values.yaml
Environment variables starting with HELM_:
HELM_DEBUG=false
"""

['Helm when Execute without values should execute with expected args and env 1']
SnapShot = """
Arguments for helm: template chartsnap charts/app1/ --namespace=default
Environment variables starting with HELM_:
HELM_DEBUG=false
"""
